### PR TITLE
Some simplication to lowering

### PR DESCRIPTION
**Pull request by sklam**
* Get rid of NumbaFunction
* Get rid of codegen.add_linking_library.
* Clean up lower_call

### CODE REVIEW

File: numba/lowering.py [hunk 1]
```diff
@@ -691,169 +691,203 @@ def lower_call(self, resty, expr):
             fnty = self.typeof(expr.func.name)
 
         if isinstance(fnty, types.ObjModeDispatcher):
```
Comment (stuartarchibald) [inline numba/lowering.py]: Am I right in thinking that these large-ish changes are a simple refactor to put the bodies of the `isinstance` branches into functions?

File: numba/targets/base.py [hunk 2]
```diff
@@ -1107,6 +1104,13 @@ def push_code_library(self, lib):
         finally:
             self._codelib_stack.pop()
 
+    def add_linking_libs(self, libs):
+        """Add list of linking librarys to the *active_code_library*.
```
Comment (stuartarchibald) [inline numba/targets/base.py]: list? just needs to be iterable?

File: numba/lowering.py (hunk 1 above)
Comment (sklam) [reply numba/lowering.py]: Yes

**Comment by stuartarchibald** [comment]
Thanks for the fixes, this looks good. CI failures are timeouts.

**Pull request merged**